gcc/rust/ast/rust-ast-builder.h

```diff
@@ -85,6 +85,10 @@ class Builder
 			       std::unique_ptr<Expr> &&tail_expr
 			       = nullptr) const;
 
+  /* Create an early return expression with an optional expression */
+  std::unique_ptr<Expr> return_expr (std::unique_ptr<Expr> &&to_return
+				     = nullptr);
```

Comment:
`std::unique_ptr<Expr> &&` I wish it was an optional instead, this would draw a clear line and we would be able to easily catch rogue null pointers.

Reply:
I think we'd need to change the AST for this then, because the `ReturnExpr` class only keeps a unique_ptr. So the builder's code would look like this:

```c++

std::unique_ptr<Expr>
Builder::return_expr (tl::optional<std::unique_ptr<Expr>> &&to_return)
{
  if (to_return)
    return std::unique_ptr<Expr> (
      new ReturnExpr (std::move (*to_return), {}, loc));

  return std::unique_ptr<Expr> (new ReturnExpr (nullptr, {}, loc));
}
```

which is quite heavy